engines/dm/inventory.cpp

```diff
@@ -0,0 +1,1137 @@
+
+	_panelContent = k0_PanelContentFoodWaterPoisoned;
+	switch (thing.getType()) {
+	case k9_ContainerThingType:
+		_panelContent = k4_PanelContentChest;
+		break;
+	case k7_ScrollThingType:
+		_panelContent = k2_PanelContentScroll;
+		break;
+	default:
+		thing = Thing::_none;
+		break;
+	}
+
+	if (thing == Thing::_none)
+		drawPanelFoodWaterPoisoned();
+	else
+		drawPanelObject(thing, false);
+}
+
+void InventoryMan::closeChest() {
+	DungeonMan &dunMan = *_vm->_dungeonMan;
+
+	bool processFirstChestSlot = true;
+	if (_openChest == Thing::_none)
+		return;
+	Container *container = (Container*)dunMan.getThingData(_openChest);
+	_openChest = Thing::_none;
+	container->getSlot() = Thing::_endOfList;
+	Thing prevThing;
+	for (int16 chestSlotIndex = 0; chestSlotIndex < 8; ++chestSlotIndex) {
+		Thing thing = _chestSlots[chestSlotIndex];
+		if (thing != Thing::_none) {
+			_chestSlots[chestSlotIndex] = Thing::_none; // CHANGE8_09_FIX
+
+			if (processFirstChestSlot) {
+				processFirstChestSlot = false;
+				*dunMan.getThingData(thing) = Thing::_endOfList.toUint16();
+				container->getSlot() = prevThing = thing;
+			} else {
+				dunMan.linkThingToList(thing, prevThing, kM1_MapXNotOnASquare, 0);
+				prevThing = thing;
+			}
+		}
+	}
+}
+
+void InventoryMan::drawPanelScrollTextLine(int16 yPos, char* text) {
+	for (char* iter = text; *iter != '\0'; ++iter) {
+		if ((*iter >= 'A') && (*iter <= 'Z'))
+			*iter -= 64;
+		else if (*iter >= '{') // this branch is CHANGE5_03_IMPROVEMENT
+			*iter -= 96;
+	}
+	_vm->_textMan->printToViewport(162 - (6 * strlen(text) / 2), yPos, k0_ColorBlack, text, k15_ColorWhite);
+}
+
+void InventoryMan::drawPanelScroll(Scroll* scroll) {
+	DisplayMan &dispMan = *_vm->_displayMan;
+
+	char stringFirstLine[300];
+	_vm->_dungeonMan->decodeText(stringFirstLine, Thing(scroll->getTextStringThingIndex()), (TextType)(k2_TextTypeScroll | k0x8000_DecodeEvenIfInvisible));
+	char *charRed = stringFirstLine;
+	while (*charRed && (*charRed != '\n'))
+		charRed++;
+
+	*charRed = '\0';
+	dispMan.blitToViewport(_vm->_displayMan->getNativeBitmapOrGraphic(k23_PanelOpenScrollIndice),
+							   _boxPanel, k72_byteWidth, k8_ColorRed, 73);
+	int16 lineCount = 1;
+	charRed++;
+	char *charGreen = charRed; // first char of the second line
+	while (*charGreen) {
+		/* BUG0_47 Graphical glitch when you open a scroll. If there is a single line of text in a scroll
+		(with no carriage return) then charGreen points to undefined data. This may result in a graphical
+		glitch and also corrupt other memory. This is not an issue in the original dungeons where all
+		scrolls contain at least one carriage return character */
+		if (*charGreen == '\n')
+			lineCount++;
+
+		charGreen++;
+	}
+
+	if (*(charGreen - 1) != '\n')
+		lineCount++;
+	else if (*(charGreen - 2) == '\n')
+		lineCount--;
+
+	int16 yPos = 92 - (7 * lineCount) / 2; // center the text vertically
+	drawPanelScrollTextLine(yPos, stringFirstLine);
+	charGreen = charRed;
+	while (*charGreen) {
```

Comment:
The original had 70 or so instead of 300 (there was no reason for me to go over 70 anyway).